File: backend/data/skill_store/imported/nanoresearch/bio-single-cell-cnv-fastcnv-r/examples/example_basic.R
#!/usr/bin/env Rscript
#' fastCNV Single-Cell Basic Analysis Example
#'
#' Demonstrates complete CNV inference workflow on single-cell data.

# =============================================================================
# Setup
# =============================================================================

# Load required packages
if (!requireNamespace("Seurat", quietly = TRUE)) {
  install.packages("Seurat")
}

library(Seurat)

# Source the fastCNV analysis module
script_dir <- dirname(sys.frame(1)$ofile)
source(file.path(script_dir, "..", "scripts", "r", "run_fastcnv.R"))

# Check if fastCNV is installed
if (!requireNamespace("fastCNV", quietly = TRUE)) {
  message("Installing fastCNV...")
  remotes::install_github("must-bioinfo/fastCNV")
}

library(fastCNV)

message("=== fastCNV Single-Cell Analysis Example ===\n")

# =============================================================================
# Step 1: Create Simulated Tumor Data
# =============================================================================

message("Step 1: Creating simulated single-cell tumor dataset...")

set.seed(42)
n_cells <- 300
n_genes <- 1000

# Create expression matrix
counts <- matrix(
  rpois(n_cells * n_genes, lambda = 5),
  nrow = n_genes,
  ncol = n_cells
)

# Gene and cell names
rownames(counts) <- paste0("GENE_", 1:n_genes)
colnames(counts) <- paste0("cell_", 1:n_cells)

# Create cell type annotations
# 40% Tumor, 20% T cells, 15% B cells, 15% Macrophages, 10% Endothelial
cell_types <- c(
  rep("Tumor", 120),
  rep("T_cell", 60),
  rep("B_cell", 45),
  rep("Macrophage", 45),
  rep("Endothelial", 30)
)

# Shuffle
cell_types <- sample(cell_types)

# Create Seurat object
seurat_obj <- CreateSeuratObject(counts = counts)
seurat_obj$cell_type <- cell_types
seurat_obj$annot <- cell_types

# Add sample metadata
seurat_obj$sample <- "Tumor1"

message(sprintf("  Created dataset: %d cells, %d genes", n_cells, n_genes))
message(sprintf("  Cell types: %s", paste(table(cell_types), collapse = ", ")))
message(sprintf("  Reference cells (T_cell, B_cell, Macrophage): %d\n",
                sum(cell_types %in% c("T_cell", "B_cell", "Macrophage"))))

# =============================================================================
# Step 2: Run fastCNV with Reference
# =============================================================================

message("Step 2: Running fastCNV with immune reference...")

result <- run_fastcnv_sc(
  seurat_obj = seurat_obj,
  sample_name = "Simulated_Tumor",
  reference_var = "annot",
  reference_label = c("T_cell", "B_cell", "Macrophage"),
  reCluster = FALSE,
  getCNVPerChromosomeArm = TRUE,
  printPlot = FALSE,
  verbose = TRUE
)

message("  fastCNV complete!\n")

# =============================================================================
# Step 3: Explore CNV Results
# =============================================================================

message("Step 3: Exploring CNV results...")

# Check CNV columns added
cnv_cols <- grep("cnv|_CNV$", colnames(result@meta.data), value = TRUE)
message(sprintf("  CNV columns added (%d):", length(cnv_cols)))
message(sprintf("    %s", paste(head(cnv_cols, 5), collapse = ", ")))
if (length(cnv_cols) > 5) {
  message(sprintf("    ... and %d more", length(cnv_cols) - 5))
}

# CNV fraction summary
if ("cnv_fraction" %in% colnames(result@meta.data)) {
  message("\n  CNV fraction summary:")
  print(summary(result@meta.data$cnv_fraction))
}

# CNV clusters
if ("cnv_clusters" %in% colnames(result@meta.data)) {
  message("\n  CNV cluster distribution:")
  cluster_table <- table(result@meta.data$cnv_clusters)
  print(cluster_table)

  # Compare with original annotations
  message("\n  Comparison: Cell Type vs CNV Clusters")
  comparison <- table(
    Cell_Type = result@meta.data$cell_type,
    CNV_Cluster = result@meta.data$cnv_clusters
  )
  print(comparison)
}

# =============================================================================
# Step 4: CNV Clustering and Classification
# =============================================================================

message("\nStep 4: Running CNV clustering and classification...")

result <- cnv_cluster(result, reference_var = "annot")
message("  CNV clustering complete.")

result <- merge_cnv_clusters(result, mergeThreshold = 0.95)
message("  Merged correlated clusters.")

result <- cnv_classification(result, cnv_thresh = 0.09)
classification_cols <- grep("_CNV_classification$", colnames(result@meta.data), value = TRUE)
message(sprintf("  Classification columns added: %s", paste(head(classification_cols, 3), collapse = ", ")))

# =============================================================================
# Step 5: Summarize by Groups
# =============================================================================

message("\nStep 5: Summarizing CNV by groups...")

# By cell type
message("\n  CNV by Cell Type:")
cell_type_summary <- summarize_cnv_by_cluster(
  result,
  group_by = "cell_type",
  metric = "cnv_fraction"
)
print(cell_type_summary)

# By CNV cluster
message("\n  CNV by CNV Cluster:")
cluster_summary <- summarize_cnv_by_cluster(
  result,
  group_by = "cnv_clusters",
  metric = "cnv_fraction"
)
print(cluster_summary)

# =============================================================================
# Step 6: Extract and Export
# =============================================================================

message("\nStep 6: Extracting and exporting results...")

# Extract CNV metadata
cnv_data <- extract_cnv_metadata(result, include_chromosomes = TRUE)
message(sprintf("  Extracted %d columns of CNV data for %d cells",
                ncol(cnv_data) - 1, nrow(cnv_data)))

# Show sample
message("\n  Sample of CNV data:")
print(head(cnv_data[, 1:min(5, ncol(cnv_data))]))

# Export
output_dir <- file.path(script_dir, "fastcnv_output")
export_cnv_results(
  result,
  output_dir = output_dir,
  prefix = "example_sc",
  export_matrix = FALSE  # Skip matrix for demo
)

message(sprintf("\n  Results exported to: %s", output_dir))

# =============================================================================
# Step 7: Visualization Examples
# =============================================================================

message("\nStep 7: Visualization examples...")

if (requireNamespace("ggplot2", quietly = TRUE)) {
  message("  Available visualizations:")
  message("    - FeaturePlot(result, features = 'cnv_fraction')")
  message("    - DimPlot(result, group.by = 'cnv_clusters')")
  message("    - VlnPlot(result, features = 'cnv_fraction')")
  message("    - plot_chr_arm_umap(result, feature = '20.p_CNV')")
} else {
  message("  Install ggplot2 for visualization: install.packages('ggplot2')")
}

message("\n  For CNV heatmap:")
message("    plot_cnv_heatmap(result, output_file = 'cnv_heatmap.pdf')")

# =============================================================================
# Summary
# =============================================================================

message("\n=== Analysis Complete ===")
message("\nKey results:")
message(sprintf("  - Total cells analyzed: %d", ncol(result)))
message(sprintf("  - CNV clusters identified: %d", length(unique(result$cnv_clusters))))
message(sprintf("  - Mean CNV fraction: %.3f", mean(result$cnv_fraction, na.rm = TRUE)))

message("\nNext steps:")
message("  1. Check exported results in:", output_dir)
message("  2. Run standard Seurat clustering and compare with CNV clusters")
message("  3. Identify tumor subclones using CNV profiles")
message("  4. Compare with known cancer CNV alterations")

message("\nFor multi-sample analysis:")
message("  results <- run_fastcnv_multi_sc(")
message("    seurat_list = list(sample1, sample2, sample3),")
message("    sample_names = c('S1', 'S2', 'S3'),")
message("    reference_var = 'annot',")
message("    reference_label = c('T_cell', 'B_cell')")
message("  )")

File: data/skill_store/imported/nanoresearch/bio-single-cell-cnv-infercnvpy/examples/example_basic.py
#!/usr/bin/env python3
"""
infercnvpy Basic Analysis Example

Demonstrates complete CNV inference workflow with simulated tumor data.
"""

import numpy as np
import pandas as pd
import scanpy as sc
import sys
from pathlib import Path

# Import analysis module
sys.path.insert(0, str(Path(__file__).parent.parent / 'scripts' / 'python'))

from infercnv_analysis import (
    run_infercnv_pipeline,
    add_gene_positions,
    cluster_by_cnv,
    identify_cnv_regions,
    summarize_cnv_by_chromosome,
    export_cnv_results
)

# ============================================================================
# Step 1: Create Simulated Tumor Data
# ============================================================================

print("=" * 60)
print("Step 1: Create Simulated Tumor Dataset")
print("=" * 60)

np.random.seed(42)
n_cells = 200
n_genes = 500

# Create expression matrix with some CNV signal
counts = np.random.poisson(5, size=(n_cells, n_genes))

# Create gene names with chromosomal positions
genes = []
chromosomes = []
starts = []
ends = []

for i in range(n_genes):
    chr_num = (i // 100) + 1  # 100 genes per chromosome
    if chr_num > 4:  # Only 4 chromosomes for this example
        chr_num = 4
    genes.append(f"GENE_{i}")
    chromosomes.append(f"chr{chr_num}")
    starts.append(i * 10000)
    ends.append((i + 1) * 10000)

# Create cell annotations
cell_types = ['T_cell'] * 40 + ['B_cell'] * 40 + ['Macrophage'] * 20 + ['Tumor'] * 100

# Create AnnData
adata = sc.AnnData(
    X=counts,
    obs=pd.DataFrame({
        'cell_type': cell_types,
    }, index=[f"cell_{i}" for i in range(n_cells)]),
    var=pd.DataFrame({
        'chromosome': chromosomes,
        'start': starts,
        'end': ends
    }, index=genes)
)

print(f"Created dataset: {adata.n_obs} cells, {adata.n_vars} genes")
print(f"Cell types: {adata.obs['cell_type'].value_counts().to_dict()}")

# ============================================================================
# Step 2: Verify Gene Positions
# ============================================================================

print("\n" + "=" * 60)
print("Step 2: Verify Gene Positions")
print("=" * 60)

print("\nGene position annotation:")
print(adata.var.head())

# Check chromosome distribution
print("\nGenes per chromosome:")
print(adata.var['chromosome'].value_counts().sort_index())

# ============================================================================
# Step 3: Run CNV Inference
# ============================================================================

print("\n" + "=" * 60)
print("Step 3: Run CNV Inference")
print("=" * 60)

# Use immune cells as reference
run_infercnv_pipeline(
    adata,
    reference_key="cell_type",
    reference_cat=["T_cell", "B_cell", "Macrophage"],
    window_size=50,
    step=5,
    key_added="cnv",
    verbose=True
)

# Check results
print("\nCNV matrix shape:", adata.obsm["X_cnv"].shape)
print("Chromosome positions:", adata.uns["cnv"]["chr_pos"])

# ============================================================================
# Step 4: Cluster by CNV Profile
# ============================================================================

print("\n" + "=" * 60)
print("Step 4: Cluster by CNV Profile")
print("=" * 60)

cluster_by_cnv(adata, key="cnv", resolution=0.5)

print("\nCNV clusters:")
print(adata.obs["cnv_leiden"].value_counts())

# Compare with original annotations
print("\nComparison of CNV clusters vs cell types:")
comparison = pd.crosstab(adata.obs["cnv_leiden"], adata.obs["cell_type"])
print(comparison)

# ============================================================================
# Step 5: Identify CNV Regions
# ============================================================================

print("\n" + "=" * 60)
print("Step 5: Identify Altered CNV Regions")
print("=" * 60)

cnv_regions = identify_cnv_regions(
    adata,
    key="cnv",
    threshold=0.2,
    min_cells=10
)

if not cnv_regions.empty:
    print("\nSignificant CNV alterations:")
    print(cnv_regions)
else:
    print("\nNo significant alterations found (expected for simulated data)")

# ============================================================================
# Step 6: Summarize by Chromosome
# ============================================================================

print("\n" + "=" * 60)
print("Step 6: Summarize CNV by Chromosome and Cell Type")
print("=" * 60)

chr_summary = summarize_cnv_by_chromosome(
    adata,
    key="cnv",
    groupby="cell_type"
)

print("\nChromosome summary by cell type:")
print(chr_summary)

# ============================================================================
# Step 7: Export Results
# ============================================================================

print("\n" + "=" * 60)
print("Step 7: Export Results")
print("=" * 60)

output_dir = Path("cnv_results")
export_cnv_results(
    adata,
    output_dir=output_dir,
    key="cnv",
    prefix="example"
)

print(f"\nResults exported to: {output_dir.absolute()}")
print("Files created:")
for f in output_dir.iterdir():
    print(f"  - {f.name}")

# ============================================================================
# Summary
# ============================================================================

print("\n" + "=" * 60)
print("Analysis Complete!")
print("=" * 60)

print("\nKey results:")
print(f"  - CNV matrix: {adata.obsm['X_cnv'].shape}")
print(f"  - CNV clusters: {adata.obs['cnv_leiden'].nunique()} clusters")
print(f"  - Output directory: {output_dir.absolute()}")

print("\nNext steps:")
print("  - Visualize with: cnv.pl.chromosome_heatmap(adata, groupby='cell_type')")
print("  - Explore tumor subclones by CNV profile")
print("  - Compare with known cancer alterations")
